feat: modelling guide (#1221)
ory/docs: docs/keto/modeling/learn-the-concepts.mdx

---
id: learn-the-concepts
title: Create your first permission model with OPL
sidebar_label: Create your first permission model
---

```mdx-code-block
import Mermaid from "@site/src/theme/Mermaid"
```

This section guides you through creating your first permission model using the Ory Permission Language (OPL).

## What is a permission model

> Permission checks are answered based on data in the database, for example Bob is owner of document X, as well as the permission
> model, for example all owners of a document can view it.
> 
> When you ask Ory Permissions if Bob can view document X, it will check if Bob is owner of the document X. The permission model
> tells Ory Permissions what to check in the database.

Reviewer: I think the permission model is rather the collection of all namespaces and permission rules. What you described is rather a single permission rule. Also, I would suggest to use the new relationship string format with syntax highlighting:

```suggestion
Permission checks are answered based on data in the database, for example <code class="language-keto-relationships">User:Bob is owner of Document:X</code>, as well as permission rules
like "all owners of a document can view it".

When you ask Ory Permissions <code class="language-keto-relationships">is User:Bob allowed to view on Document:X</code>, it will check if Bob is owner of the document X. The permission model
tells Ory Permissions what to check in the database.
```

Reviewer: btw I figured out the inline syntax highlighting does not work...

## Define a permission model

Designing a permission model is a complex task. Ory Permissions and the Ory Permission Language, as a subset of TypeScript,
provide a streamlined approach to permissions with the benefit of being processed by a fast, global permission engine.

Just as there is no single approach for programming, there is no universally applicable guide for constructing a permission model.
Nevertheless, the following iterative process can be a good starting point:

> 1. Create a list of objects that you want to give access to.

Reviewer: ```suggestion
1. Create a list of objects that you want to control access to.
1. Create a list of subjects (actors) that need to have access to objects.
```

1. For each object

>    1. Create a list of relationships it has with other objects. In a database, these would be associations expressed with foreign
>       keys.

Reviewer: ```suggestion
   1. Create a list of relationships it has with other objects. You can think of these as associations expressed as foreign keys in classical relational databases.```

   1. Create a list of permissions that you want to check.
1. Define each relationship in the Ory Permission Language (OPL).

> 1. Define each permission in the OPL.

Reviewer: ```suggestion
1. Define each permission in the OPL. Permissions are derived from relationships.
```

1. Test your permission model.

## Example: document store

To guide you through the process of defining a permission model, this example will be used:

> We want to model user access to documents such that users can be an owner, editor, or viewer of a document. An owner is also an
> editor of a document, and an editor is also a viewer of a document. Furthermore, documents can be put into a hierarchy of
> folders, such that being able to view the parent folder implies being able to view the enclosed folders and documents.

> ### Create a list of objects
> 
> We start with the description and highlight all objects that we find:
> 
> > We want to model `user` access to `documents` such that `users` can be an owner, editor, or viewer of a document. An owner is
> > also an editor of a `document`, and an editor is also a viewer of a `document`. Furthermore, `documents` can put into a
> > hierarchy of `folders`, where the ability to view the parent `folder` implies the ability to view the enclosed `folders` and
> > `documents`.
> 
> We found these objects:
> 
> - User
> - Document
> - Folder

Reviewer: Somehow it looks weird when we talk about users as objects.

```suggestion
### Create a list of objects and subjects

We start with the description and highlight all **objects** and _subjects_ that we find:

> We want to model _user_ access to **documents** such that _users_ can be an owner, editor, or viewer of a document. An owner is
> also an editor of a **document**, and an editor is also a viewer of a **document**. Furthermore, **documents** can put into a
> hierarchy of **directories**, where the ability to view the parent **directory** implies the ability to view the enclosed **directories** and
> **documents**.

Therefore, we have the `User` subject and `Document` and `Directory` objects.
```

see [folders vs directories](https://www.baeldung.com/cs/directories-vs-folders).

This list translates into our first version of the permission model.

```ts title="permissions-v1.ts"
import { Namespace, Context } from "@ory/keto-namespace-types"

// highlight-start
class User implements Namespace {}
class Document implements Namespace {}
class Folder implements Namespace {}
// highlight-end
```

:::tip

The objects and subjects from your description are [namespaces](../concepts/namespaces) in Ory Permissions, and are declared as
classes in the Ory Permission Language. The convention is to spell these class names as TypeScript classes, capitalized and in
singular form, for example `Document` instead of `documents`.

:::

> ### List relationships to other objects

Reviewer: ```suggestion
### Determine relationships between objects
```

Read through the description sentence by sentence to find the list of relationships:

> We want to model `user` access to `documents` such that `users` can be an `owner`, `editor`, or `viewer` of a `document`.

<Mermaid
  chart={`
graph LR
    Document -->|owner| User
    Document -->|editor| User
    Document -->|viewer| User
`}
/>

> An owner is also an editor of a `document`, and an editor is also a viewer of a `document`. Furthermore, `documents` can put
> into a hierarchy of `folders`, such that being able to view the parent `folder` implies being able to view the enclosed
> `folders` and `documents`.

<Mermaid
  chart={`
graph LR
    Document -->|parent| Folder
    Folder -->|parent| Folder
    Folder -->|owner| User
    Folder -->|editor| User
    Folder -->|viewer| User
`}
/>

Putting this all together in one graph:

<Mermaid
  chart={`
graph LR
    Document -->|owner| User
    Document -->|editor| User
    Document -->|viewer| User
    Document -->|parent| Folder
    Folder -->|parent| Folder
    Folder -->|owner| User
    Folder -->|editor| User
    Folder -->|viewer| User
`}
/>

### List permissions to check for each object

> When you perform an action on behalf of a user, you usually check a specific permission, such as `view` or `edit`, instead of a
> relationship, such as `owner`, and have the relationship implied by the permission.

Reviewer: ```suggestion
When you perform an action on behalf of a user, you should check a specific permission, such as `view` or `edit`, instead of a
relationship, such as `owner`. Therefore, each permission is derived from relationships and other permissions.
```

> When you perform an action on behalf of a user, you usually check a specific permission, such as `view` or `edit`, instead of a
> relationship, such as `owner`, and have the relationship implied by the permission.

Reviewer 2: ```suggestion
When you perform an action on behalf of a user, you often want to know what permissions the user has, for example `view` or `edit` rather than the relationship, for example `owner`. In this case you check the relationship of the user, which then implies the permissions this user has.
```
I misunderstood this the first time around, this should be more clear, wdyt

Author: I applied Patrik's suggestion, does this make things clearer already?

> Add these permissions to the database to express what the application needs. For our document storage, you have these permissions:

Reviewer: I don't get the first sentence. Add them to which database? How do I add it?

> Add these permissions to the database to express what the application needs. For our document storage, you have these permissions:

Author: ```suggestion
Add these permissions to the model to express what the application needs. For our document storage, you have these permissions:
```

- `view` a document if the user is a `viewer`, `editor`, or `owner` of the document; or if the user can `view` the parent folder
- `edit` a document if the user is a `editor`, or `owner` of the document; or if the user can `edit` the parent folder
- `delete` a document if the user is an `owner` of the document; or if the user can `delete` the parent folder
- `share` a document if the user is an `owner` of the document; or if the user can `share` the parent folder
- `delete` a folder if the user is an `owner` of the folder; or if the user can `delete` the parent folder
- `share` a folder if the user is an `owner` of the folder; or if the user can `share` the parent folder

> ### Define relations in the OPL

Reviewer: I think this should happen before the permission topic starts.

Author: You mean instead of

1. Create a list of objects, which are things that you want to give access to.
2. For each object
   1. Make a list of relationships this has to other objects. In a database, those would be associatons expressed with foreign keys.
   2. Make a list of permissions that you want to check.
5. Define each relation in the OPL
6. Define each permission in the OPL
7. Test your permission model

the outline would be

1. Create a list of objects, which are things that you want to give access to.
1. Make a list of relationships this has to other objects. In a database, those would be associatons expressed with foreign keys.
5. Define each relation in the OPL
3. Make a list of permissions that you want to check.
6. Define each permission in the OPL
7. Test your permission model

?

Reviewer: I mean because you first determine the relationships, then you should directly add them to the OPL before you go into the permissions topic.

Author: agreed

Next, add each relation to the permission config.

:::tip

> The relations from the description translate to relations in Ory Permissions, which are declared inside the corresponding class in

Reviewer: I am not sure which description this is referring to.
```suggestion
The relations from the description translate to relationships in Ory Permissions, which are declared inside the corresponding class in
```

the Ory Permission Language. Since you store arbitrary tuples in Ory Permissions, you always have many-to-many relations between
these objects. To reflect this in our language, the convention is to pluralize the relation name, i.e., `viewers` instead of
`viewer`.

:::

```ts title="permissions-v2.ts"
import { Namespace, Context } from "@ory/keto-namespace-types"

class User implements Namespace {}

class Document implements Namespace {
  // highlight-start
  related: {
    owners: User[]
    editors: User[]
    viewers: User[]
    parents: Folder[]
  }
  // highlight-end
}

class Folder implements Namespace {
  // highlight-start
  related: {
    owners: User[]
    editors: User[]
    viewers: User[]
    parents: Folder[]
  }
  // highlight-end
}
```

### Define permissions in the OPL

The permissions are expressed in the OPL as TypeScript functions that take a context containing the subject, for example the

> current user, and that answer the permission check based on the relations this object has, in this case the current document.

Reviewer: ```suggestion
current user, and that answer the permission check based on the relationships this object has, in this case the current document.
```

    //highlight-end
  }
}

class Folder implements Namespace {
  related: {
    owners: User[]
    editors: User[]
    viewers: User[]
    parents: Folder[]
  }
}
```

Let's do this for the remaining permissions:

```ts title="permissions-v4.ts"
import { Namespace, Context } from "@ory/keto-namespace-types"

class User implements Namespace {}

class Document implements Namespace {
  related: {
    owners: User[]
    editors: User[]
    viewers: User[]
    parents: Folder[]
  }

  permits = {
    view: (ctx: Context): boolean =>
      this.related.viewers.includes(ctx.subject) ||
      this.related.editors.includes(ctx.subject) ||
      this.related.owners.includes(ctx.subject) ||
      this.related.parents.traverse((parent) => parent.permits.view(ctx)),
    //highlight-start
    edit: (ctx: Context): boolean =>
      this.related.editors.includes(ctx.subject) ||
      this.related.owners.includes(ctx.subject) ||
      this.related.parents.traverse((parent) => parent.permits.edit(ctx)),
    delete: (ctx: Context): boolean =>
      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.delete(ctx)),
    share: (ctx: Context): boolean =>

>       this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),

Reviewer: In this case we have kind of hierarchical permissions, i.e. `delete = share > edit > view`. IMO it would be best practice to implement this hierarchy also in the permissions, as that ensures the hierarchy is always preserved. Therefore, permission escalations due to misconfiugration are less likely.

This however is only the case for hierarchical permissions, so it depends on the use-case if that is valid. Should we do it here or just keep as is?

Author: Hmm, I was also thinking about a section to refactor the permission model, showing the benefit of it being "mostly TypeScript". Then you would write

```ts
  permits = {
￼    view: (ctx: Context): boolean =>
￼      this.related.viewers.includes(ctx.subject) ||
￼      this.permits.edit(ctx),
￼  // ...
```
and so on.

Reviewer: Yes exactly. This has the advantage that you automatically include one permission in the other, which makes it less likely that there is an error. But that only holds for cases like this where you can actually do that.

```

This section guides you through creating your first permission model using the Ory Permission Language (OPL).

## What is a permission model

Permission checks are answered based on data in the database, for example <code class="language-keto-relationships">User:Bob is
owner of Document:X</code>, as well as permission rules like "all owners of a document can view it".

When you ask Ory Permissions <code class="language-keto-relationships">is User:Bob allowed to view on Document:X</code>, it will
look up how Bob could have the `view` permission, and then check if Bob is owner of the document X. The permission model tells Ory
Permissions what to check in the database.

## Define a permission model

Designing a permission model is a complex task. Ory Permissions and the Ory Permission Language, as a subset of TypeScript,
provide a streamlined approach to permissions with the benefit of being processed by a fast, global permission engine.

Just as there is no single approach for programming, there is no universally applicable guide for constructing a permission model.
Nevertheless, the following iterative process can be a good starting point:

> 1. Create a list of objects, which are things that you want to give access to.

Reviewer 3: Do you always want to give access to objects? 
```suggestion
1. Create a list of objects. Objects are the entities that you want to manage access for.
```

Reviewer 3: Link to objects doc? 
https://ory.sh/docs/keto/concepts/objects

    Document -->|editor| User
    Document -->|viewer| User
`}
/>

> An owner is also an editor of a `document`, and an editor is also a viewer of a `document`. Furthermore, `documents` can put
> into a hierarchy of `folders`, such that being able to view the parent `folder` implies being able to view the enclosed
> `folders` and `documents`.

<Mermaid
  chart={`
graph LR
    Document -->|parent| Folder
    Folder -->|parent| Folder
    Folder -->|owner| User
    Folder -->|editor| User
    Folder -->|viewer| User
`}
/>

Putting this all together in one graph:

<Mermaid
  chart={`
graph LR
    Document -->|owner| User
    Document -->|editor| User
    Document -->|viewer| User
    Document -->|parent| Folder
    Folder -->|parent| Folder
    Folder -->|owner| User
    Folder -->|editor| User
    Folder -->|viewer| User
`}
/>

### Define relations in the OPL

Next, add each relation to the permission config.

:::tip

The relations from the description translate to relationships in Ory Permissions, which are declared inside the corresponding
class in the Ory Permission Language. Since you store arbitrary tuples in Ory Permissions, you always have many-to-many relations

> between these objects. To reflect this in our language, the convention is to pluralize the relation name, i.e., `viewers` instead

Reviewer 3: > To reflect this in our language

Do you mean English? German? Polish? :p

> between these objects. To reflect this in our language, the convention is to pluralize the relation name, i.e., `viewers` instead

Reviewer 3: ```suggestion
between these objects. To reflect this in our language, the convention is to pluralize the relation name, for example, `viewers` instead
```

relationship, such as `owner`. Therefore, each permission is derived from relationships and other permissions.

Add these permissions to the model to express what the application needs. For our document storage, you have these permissions:

- `view` a document if the user is a `viewer`, `editor`, or `owner` of the document; or if the user can `view` the parent folder
- `edit` a document if the user is a `editor`, or `owner` of the document; or if the user can `edit` the parent folder
- `delete` a document if the user is an `owner` of the document; or if the user can `delete` the parent folder
- `share` a document if the user is an `owner` of the document; or if the user can `share` the parent folder
- `delete` a folder if the user is an `owner` of the folder; or if the user can `delete` the parent folder
- `share` a folder if the user is an `owner` of the folder; or if the user can `share` the parent folder

### Define permissions in the OPL

The permissions are expressed in the OPL as TypeScript functions that take a context containing the subject, for example the

> current user, and that answer the permission check based on the relationships this object has, in this case the current document.

Reviewer 3: > on the relationships this object has, in this case the current document.

Is the "current document" a relationship?

Author: No, current document would be the `this`

:::tip

The permissions from the description are declared as functions inside the `permits` property of the corresponding class in the Ory

> Permission Language.

Reviewer 3: You introduced the OPL contraction, it's okay to use it instead of the full name.

:::

Let's see this in action for the first permission:

> - `view` a document if the user is a `viewer`, `editor`, or `owner` of the document; or if the user can `view` the parent folder

Reviewer 3: Wouldn't this be clearer if split into two "bullets"?
```suggestion
- `view` the document if the user is a `viewer`, `editor`, or `owner` of the document  
- `view` the document if the user can `view` the parent folder
```

      this.related.parents.traverse((parent) => parent.permits.view(ctx)),
    //highlight-start
    edit: (ctx: Context): boolean =>
      this.related.editors.includes(ctx.subject) ||
      this.related.owners.includes(ctx.subject) ||
      this.related.parents.traverse((parent) => parent.permits.edit(ctx)),
    delete: (ctx: Context): boolean =>
      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.delete(ctx)),
    share: (ctx: Context): boolean =>
      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),
    //highlight-end
  }
}

class Folder implements Namespace {
  related: {
    owners: User[]
    editors: User[]
    viewers: User[]
    parents: Folder[]
  }

  //highlight-start
  permits = {
    delete: (ctx: Context): boolean =>
      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.delete(ctx)),
    share: (ctx: Context): boolean =>
      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),
  }
  //highlight-end
}
```

> ### Optional: Refactor your permission model code

Reviewer 3: ```suggestion
### Optional: refactor permission model code
```

      this.related.editors.includes(ctx.subject) ||
      //highlight-next-line
      this.permits.share(ctx),
    delete: (ctx: Context): boolean =>
      //highlight-next-line
      this.permits.share(ctx),
    share: (ctx: Context): boolean =>
      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),
  }
}

class Folder implements Namespace {
  related: {
    owners: User[]
    editors: User[]
    viewers: User[]
    parents: Folder[]
  }

  permits = {
    delete: (ctx: Context): boolean =>
      //highlight-next-line
      this.permits.share(ctx),
    share: (ctx: Context): boolean =>
      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),
  }
}
```

Whether or not this refatoring makes sense in your application depends on your requirements, of course. The take-away is that
these kind of refacorings are possible because the permission config is essentially just code. So you can apply the same techiques
your permission config as to your code to create well-structured and maintainable software systems.

### Test permissions

> It is important to test your permission model. To explore the permissions, you can

Reviewer 3: You say it's important to test. Why are you talking about "exploring" then?

Author: I meant "test manually" here

[create relationships and check permissions](../guides/simple-access-check-guide.mdx) through the API or SDK.

For continuous testing, we recommend the following best practices:

> - Automate testing your permission model. Write a test, in a language of your choice, that inserts the relationships and checks
>   the permissions through the [SDK](../sdk/01_overview.md).

Reviewer 3: ```suggestion
- Automate testing your permission model. Write a test, that inserts the relationships and checks
  the permissions through the [SDK](../sdk/01_overview.md). Use your preferred programming language.
```

- For complex permission model changes, use a separate Ory Network project. Each Ory Network project has an isolated permission

>   model, so you can iterate on and test your changes on a test project and only deploy the changes once all tests pass.

Reviewer 3: > Each Ory Network project has an isolated permission model

This document talks about CREATING a permission model. Does every project have one by default? Or did you mean that every project is an isolated env. that can have its own permission model?

Author: I meant that every that each project has its own (isolated) permission model.
